### PR TITLE
Feat: Find the pids of overflowd subtransaction

### PR DESCRIPTION
**Pull request by jpzhu-edu**
Its function is the same as #13539
In gpdb, it supports using `savepoint`, `begin...exception...`, and `plpython`
to issue the subtransactions. Especially, it may be much more serious for the
application error handling. It may issue many more subtransactions for
error handling.

For the postgres and gpdb, the number of active subtransactions has a maximum
value. The subtransaction state moves back and force when the number of
active subtransactions exceeds the maximum value (64) to inspect the
snapshot visibility. It may cause a shake in system performance.

The DBA needs to inspect the cause. So, we add the helper function for
finding the pids of overflowed subtransaction for coordinator and segments.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
Code change LGTM.

1. why we need the 3rd test case?
2. please add comments on test cases to demonstrate the object.

**Comment by jpzhu-edu** [comment]
> Code change LGTM.
> 
> 1. why we need the 3rd test case?
> 2. please add comments on test cases to demonstrate the object.

Sorry for the late response!  
1. The corner case is not a subtransaction overflow. I test the code is correct.
2. I will add the comment lately.

**Comment by jpzhu-edu** [comment]
```c
Datum
gp_subtrx_overflow_pids(PG_FUNCTION_ARGS)
{
	FuncCallContext *funcctx = NULL;
	SubTrxContext 	*fctx = NULL;	/* User function context. */

	if (SRF_IS_FIRSTCALL())
	{
		TupleDesc	tupdesc;
		MemoryContext oldcontext;

		funcctx = SRF_FIRSTCALL_INIT();

		/* Switch context when allocating stuff to be used in later calls */
		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);

		/* build tuple descriptor */
		tupdesc = CreateTemplateTupleDesc(2);

		TupleDescInitEntry(tupdesc, (AttrNumber) 1, "segid", INT4OID, -1, 0);
		TupleDescInitEntry(tupdesc, (AttrNumber) 2, "pids", TEXTOID, -1, 0);

		funcctx->tuple_desc = BlessTupleDesc(tupdesc);

		fctx = (SubTrxContext *) palloc(sizeof(SubTrxContext));
		fctx->cdb_pgresults.pg_results = NULL;
		fctx->cdb_pgresults.numResults = 0;
		fctx->index = 0;

		funcctx->user_fctx = (void *) fctx;
		if(Gp_role == GP_ROLE_DISPATCH)
			CdbDispatchCommand("SELECT * FROM pg_catalog.gp_subtrx_overflow_pids()",
						   		DF_NONE, &fctx->cdb_pgresults);

		fctx->record = gp_subtx_internal_helper();
		funcctx->user_fctx = (void *)fctx;
		MemoryContextSwitchTo(oldcontext);
	}

	/*
	 * Using SRF to return all the segment information of the form
	 * {segid, pids}
	 */
	funcctx = SRF_PERCALL_SETUP();

	/* Get the saved state. */
	fctx = funcctx->user_fctx;

	while (fctx->index <= fctx->cdb_pgresults.numResults)
	{
		Datum		values[2];
		bool		nulls[2];
		HeapTuple	tuple;
		Datum		result;
		char		*pids;
		int			seg_index;
		if (fctx->index == 0)
		{
			/* Setting fields representing QD's result */
			seg_index = GpIdentity.segindex;
			pids = fctx->record->pids;
		}
		else
		{
			/* Setting fields representing QE's result */
			seg_index = fctx->index - 1;
			struct pg_result *pgresult = fctx->cdb_pgresults.pg_results[seg_index];
			for (int j = 0; j < PQntuples(pgresult); j++)
			{
				pids = PQgetvalue(pgresult, j, 1);
			}
		}

		/*
		 * Form tuple with appropriate data.
		 */
		MemSet(values, 0, sizeof(values));
		MemSet(nulls, false, sizeof(nulls));

		values[0] = Int32GetDatum(seg_index);
		values[1] = CStringGetTextDatum(pids);
		tuple = heap_form_tuple(funcctx->tuple_desc, values, nulls);
		result = HeapTupleGetDatum(tuple);

		fctx->index++;
		SRF_RETURN_NEXT(funcctx, result);
	}

	SRF_RETURN_DONE(funcctx);
}

static SubTrxRec*
gp_subtx_internal_helper()
{
	int				i;
	StringInfoData 	buf;
	bool			traverse = false;
	SubTrxRec		*result;

	result = (SubTrxRec*) palloc(sizeof(SubTrxRec));

	initStringInfo(&buf);

	LWLockAcquire(ProcArrayLock, LW_EXCLUSIVE);

	for (i = 0; i < ProcGlobal->allProcCount; i++)
	{
		if (ProcGlobal->allPgXact[i].overflowed)
		{
			if (traverse)
				appendStringInfoString(&buf, ",");
			appendStringInfo(&buf, "%d", ProcGlobal->allProcs[i].pid);
			traverse = true;
		}
	}
	result->segid = GpIdentity.segindex;
	result->pids = buf.data;
	LWLockRelease(ProcArrayLock);

	return result;
}
```

On the following Monday, I will submit a new pr. It is optional to replace this pr.

**Comment by jpzhu-edu** [comment]
> ```c
> Datum
> gp_subtrx_overflow_pids(PG_FUNCTION_ARGS)
> {
> 	FuncCallContext *funcctx = NULL;
> 	SubTrxContext 	*fctx = NULL;	/* User function context. */
> 
> 	if (SRF_IS_FIRSTCALL())
> 	{
> 		TupleDesc	tupdesc;
> 		MemoryContext oldcontext;
> 
> 		funcctx = SRF_FIRSTCALL_INIT();
> 
> 		/* Switch context when allocating stuff to be used in later calls */
> 		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
> 
> 		/* build tuple descriptor */
> 		tupdesc = CreateTemplateTupleDesc(2);
> 
> 		TupleDescInitEntry(tupdesc, (AttrNumber) 1, "segid", INT4OID, -1, 0);
> 		TupleDescInitEntry(tupdesc, (AttrNumber) 2, "pids", TEXTOID, -1, 0);
> 
> 		funcctx->tuple_desc = BlessTupleDesc(tupdesc);
> 
> 		fctx = (SubTrxContext *) palloc(sizeof(SubTrxContext));
> 		fctx->cdb_pgresults.pg_results = NULL;
> 		fctx->cdb_pgresults.numResults = 0;
> 		fctx->index = 0;
> 
> 		funcctx->user_fctx = (void *) fctx;
> 		if(Gp_role == GP_ROLE_DISPATCH)
> 			CdbDispatchCommand("SELECT * FROM pg_catalog.gp_subtrx_overflow_pids()",
> 						   		DF_NONE, &fctx->cdb_pgresults);
> 
> 		fctx->record = gp_subtx_internal_helper();
> 		funcctx->user_fctx = (void *)fctx;
> 		MemoryContextSwitchTo(oldcontext);
> 	}
> 
> 	/*
> 	 * Using SRF to return all the segment information of the form
> 	 * {segid, pids}
> 	 */
> 	funcctx = SRF_PERCALL_SETUP();
> 
> 	/* Get the saved state. */
> 	fctx = funcctx->user_fctx;
> 
> 	while (fctx->index <= fctx->cdb_pgresults.numResults)
> 	{
> 		Datum		values[2];
> 		bool		nulls[2];
> 		HeapTuple	tuple;
> 		Datum		result;
> 		char		*pids;
> 		int			seg_index;
> 		if (fctx->index == 0)
> 		{
> 			/* Setting fields representing QD's result */
> 			seg_index = GpIdentity.segindex;
> 			pids = fctx->record->pids;
> 		}
> 		else
> 		{
> 			/* Setting fields representing QE's result */
> 			seg_index = fctx->index - 1;
> 			struct pg_result *pgresult = fctx->cdb_pgresults.pg_results[seg_index];
> 			for (int j = 0; j < PQntuples(pgresult); j++)
> 			{
> 				pids = PQgetvalue(pgresult, j, 1);
> 			}
> 		}
> 
> 		/*
> 		 * Form tuple with appropriate data.
> 		 */
> 		MemSet(values, 0, sizeof(values));
> 		MemSet(nulls, false, sizeof(nulls));
> 
> 		values[0] = Int32GetDatum(seg_index);
> 		values[1] = CStringGetTextDatum(pids);
> 		tuple = heap_form_tuple(funcctx->tuple_desc, values, nulls);
> 		result = HeapTupleGetDatum(tuple);
> 
> 		fctx->index++;
> 		SRF_RETURN_NEXT(funcctx, result);
> 	}
> 
> 	SRF_RETURN_DONE(funcctx);
> }
> 
> static SubTrxRec*
> gp_subtx_internal_helper()
> {
> 	int				i;
> 	StringInfoData 	buf;
> 	bool			traverse = false;
> 	SubTrxRec		*result;
> 
> 	result = (SubTrxRec*) palloc(sizeof(SubTrxRec));
> 
> 	initStringInfo(&buf);
> 
> 	LWLockAcquire(ProcArrayLock, LW_EXCLUSIVE);
> 
> 	for (i = 0; i < ProcGlobal->allProcCount; i++)
> 	{
> 		if (ProcGlobal->allPgXact[i].overflowed)
> 		{
> 			if (traverse)
> 				appendStringInfoString(&buf, ",");
> 			appendStringInfo(&buf, "%d", ProcGlobal->allProcs[i].pid);
> 			traverse = true;
> 		}
> 	}
> 	result->segid = GpIdentity.segindex;
> 	result->pids = buf.data;
> 	LWLockRelease(ProcArrayLock);
> 
> 	return result;
> }
> ```
> 
> On the following Monday, I will submit a new pr. It is optional to replace this pr.

I have not test many more cases. Next Monday.

**Comment by jpzhu-edu** [comment]
```sql
t=# select * from gp_subtrx_overflow_pids();
 segid | pids
-------+-------
    -1 | 56110
     0 | 56112
     1 | 56113
     2 | 56114
(4 rows)
```

**Comment by ashwinstar** [comment]
Would it be better and more useful to provide `mppSessionId` instead of pid? (specially in-case of multi-slice queries the output will unnecessarily report many PIDs for same session - all PIDs of same session should have overflowed at same time)?

I am guessing we do have to backport this functionality to 6X_STABLE. What's the plan for that - as an extension?

File: src/backend/cdb/cdbutil.c [hunk 1]
```diff
@@ -1800,3 +1802,29 @@ AvoidCorefileGeneration()
 	}
 #endif
 }
+
+/*
+ * Find the pids when subtransaction overflowed.
+ */
+Datum
+gp_find_subtx_overflowed_pids(PG_FUNCTION_ARGS)
```
Comment (soumyadeep2007) [inline src/backend/cdb/cdbutil.c:1810]: This should return an ARRAY to make it more application friendly. Otherwise, the application will have to write parsing logic and helps avoid trailing commas like:

```
postgres=# select gp_find_subtx_overflowed_pids();
 gp_find_subtx_overflowed_pids 
-------------------------------
 {1320857, }
(1 row)
```

File: src/test/regress/expected/subtrx_overflow.out [hunk 2]
```diff
@@ -0,0 +1,97 @@
+drop table if exists t_13524_1;
```
Comment (soumyadeep2007) [inline src/test/regress/expected/subtrx_overflow.out:1]: The test cases need comments.

**Review by soumyadeep2007 — COMMENTED** [review]
I think we need to think carefully about the design and implementation of
this feature. And we can choose to be more crafty in 7X vs 6X.


1. What we can afford in 7X is to introduce a suboverflowed field to the
pg_stat_activity view. This means adding a boolean to PgBackendStatus, update
pg_stat_get_activity() etc.
	```
	GetNewTransactionId()
	{
		...
	    int            nxids = MyPgXact->nxids;
	    if (nxids < PGPROC_MAX_CACHED_SUBXIDS)
	    {
	        MyProc->subxids.xids[nxids] = xid;
	        pg_write_barrier();
	        MyPgXact->nxids = nxids + 1;
	    }
	    else
	        MyPgXact->overflowed = true;
	        pg_stat_report_suboverflowed(true); // will update PgBackendStatus->suboverflowed
	    ...
	}

	and 

	GXactLoadSubxactData()
	{
		...
		if (nsubxacts > PGPROC_MAX_CACHED_SUBXIDS)
		{
			pgxact->overflowed = true;
			pg_stat_report_suboverflowed(true); // will update PgBackendStatus->suboverflowed
			nsubxacts = PGPROC_MAX_CACHED_SUBXIDS;
		}
		...
	```
	And then in `Abort|Commit|PrepareTransaction()` we can call 
`pg_stat_report_suboverflowed(false)`
Note that this approach avoids taking the `ProcArrayLock`, making it more
performant.

2. I think it is high time we introduce the system view: `gp_stat_activity`
which can be defined as:
`select gp_execution_segment(), * from gp_dist_random('pg_stat_activity');`

3. Now if anyone wishing to find out if there are overflown subtxs or not in the
system, one can simply run:
`select * from gp_stat_activity where suboverflowed = true group by session_id order by gp_segment_id;`
We want to make this query ^ into a system view:
`gp_suboverflowed_backends`

4. We might want to also consider adding a log message to print the query string
that caused the overflow. This is important as only 1 statement out of thousands
executed in a backend may trigger the overflow, or the backend can come out of
the overflow state before it is inspected with our view/UDF. Logging the
statement will ensure that customers can pinpoint the offending queries. We
can further guard it by a guc: `gp_log_suboverflow_statement`

5. For 6X, your current implementation should be fine as we can't really modify
pg_stat_activity. However, we should prefer a user facing view, like `gp_suboverflowed_backends` (backed by your
UDF)

File: src/backend/cdb/cdbutil.c (hunk 1 above)
Comment (jpzhu-edu) [reply src/backend/cdb/cdbutil.c:1810]: > This should return an ARRAY to make it more application friendly. Otherwise, the application will have to write parsing logic and helps avoid trailing commas like:
> 
> ```
> postgres=# select gp_find_subtx_overflowed_pids();
>  gp_find_subtx_overflowed_pids 
> -------------------------------
>  {1320857, }
> (1 row)
> ```

If the return value of UDF is set of record, the array as return value maybe not be allowed.
And I will change the output result of `cstring` type.

File: src/test/regress/expected/subtrx_overflow.out (hunk 2 above)
Comment (jpzhu-edu) [reply src/test/regress/expected/subtrx_overflow.out:1]: > The test cases need comments.

I will add the comment.

**Comment by jpzhu-edu** [comment]
> Would it be better and more useful to provide `mppSessionId` instead of pid? (specially in-case of multi-slice queries the output will unnecessarily report many PIDs for same session - all PIDs of same session should have overflowed at same time)?
> 
> I am guessing we do have to backport this functionality to 6X_STABLE. What's the plan for that - as an extension?

Yes, Ashwin. In the 6X_STABLE, we will use the extension. we can't do catalog changes for a stable version.

**Comment by jpzhu-edu** [comment]
Sorry, I forgot to rebase repo, which leads to corrupt commit history. So, I close the pr and reopen #13554.
